### PR TITLE
[MLIR][LLVM] Add fast-math related function attribute support

### PR DESCRIPTION
**Pull request by asb**
Adds unsafe-fp-math, no-infs-fp-math, no-nans-fp-math, approx-func-fp-math, and no-signed-zeros-fp-math function attributes.

This allows code generators using the LLVMIR dialect to match the codegen of Clang.

### CODE REVIEW

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td [hunk 1]
```diff
@@ -1428,7 +1428,12 @@ def LLVM_LLVMFuncOp : LLVM_Op<"func", [
     OptionalAttr<LLVM_VScaleRangeAttr>:$vscale_range,
     OptionalAttr<FramePointerKindAttr>:$frame_pointer,
     OptionalAttr<StrAttr>:$target_cpu,
-    OptionalAttr<LLVM_TargetFeaturesAttr>:$target_features
+    OptionalAttr<LLVM_TargetFeaturesAttr>:$target_features,
+    OptionalAttr<BoolAttr>:$unsafe_fp_math,
+    OptionalAttr<BoolAttr>:$no_infs_fp_math,
+    OptionalAttr<BoolAttr>:$no_nans_fp_math,
+    OptionalAttr<BoolAttr>:$approx_func_fp_math,
+    OptionalAttr<BoolAttr>:$no_signed_zeros_fp_math
```
Comment (zero9178) [inline mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:1436]: I am sadly not too familier with these attributes, and documentation seems very sparse, so I am curious whether these attributes are best modelled as `OptionalAttr<BoolAttr>` or `OptionalAttr<UnitAttr>`.

The former is tri-state between "not present", "true" and "false", but does LLVM actually differentiate between "not-present" and "false" anywhere? If not, making it a `OptionalAttr<UnitAttr>` would probably be cleaner.

File: mlir/test/Target/LLVMIR/Import/function-attributes.ll
Comment (zero9178) [inline mlir/test/Target/LLVMIR/Import/function-attributes.ll:1]: If we keep the `BoolAttr`, I think it'd be worth adding at least one test where the value is "false" as well. Ditto in the export

**Review by zero9178 — COMMENTED** [review]
Thank you so much for working on this! Code itself looks good to me although I am curious about the `OptionalAttr<BoolAttr>`. 

Is there any documentation one can read up on these attributes? They do not seem to be present in the language reference, requiring to resort to reading LLVM code to try and deduce semantics it seems.

File: mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td (hunk 1 above)
Comment (asb) [reply mlir/include/mlir/Dialect/LLVMIR/LLVMOps.td:1436]: That's a great question. As you say, documentation is unfortunately very sparse. I decided that closely matching the representation of the LLVM IR attribute may be less confusing and less prone to error, but I could be wrong.

It looks like the case you'd want to have an explicit false is for interaction with the `-enable-unsafe-fp-math` flag. `llvm/test/CodeGen/X86/change-unsafe-fp-math.ll` appears to test this interaction, where the `"unsafe-fp-math="false"` attribute does indeed override the codegen flag. So my conclusion is that BoolAttr is probably needed.

File: mlir/test/Target/LLVMIR/Import/function-attributes.ll
Comment (asb) [reply mlir/test/Target/LLVMIR/Import/function-attributes.ll:1]: I've pushed a commit that adds tests for "false" for all the added attributes in this file and fp-math-function-attributes.mlir

**Review by zero9178 — APPROVED** [review]
LGTM

**Review by gysit — APPROVED** [review]
Thanks LGTM!

**Pull request merged**